8267989: Exceptions thrown during upcalls should be handled

JornVernee opened the pull request:
Hi,

This patch regularizes exception handling for exceptions thrown during upcalls. Exceptions thrown during upcalls are now always handled by printing out the stack trace and then calling `System::exit` (see the JBS issue for some motivation).

I've added some documentation for the exception handling to `CLinker::upcallStub`, as well as a new public `int` constant in `CLinker` which is the error code that is passed to `System::exit`. The returned error code can also be configured by a system property, which for now is mostly useful for testing purposes to make sure we don't get a consistent false positive.

Thanks,
Jorn

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issues
 * [JDK-8267989](https://bugs.openjdk.java.net/browse/JDK-8267989): Exceptions thrown during upcalls should be handled
 * [JDK-8268031](https://bugs.openjdk.java.net/browse/JDK-8268031): VarHandle combinator check for exceptions is too strict


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/543/head:pull/543` \
`$ git checkout pull/543`

Update a local copy of the PR: \
`$ git checkout pull/543` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/543/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 543`

View PR using the GUI difftool: \
`$ git pr show -t 543`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/543.diff">https://git.openjdk.java.net/panama-foreign/pull/543.diff</a>

</details>

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -224,9 +232,14 @@ static SymbolLookup systemLookup() {
+     * The returned memory address is associated with the provided scope. When such scope is closed,
      * the corresponding native stub will be deallocated.
      * <p>
+     * Any exceptions that occur during an upcall should be handled during the upcall. The target method handle
+     * should not throw any exceptions. If the target method handle does throw an exception, it will be handle by
+     * calling {@link System#exit System.exit(ERR_UNCAUGHT_EXCEPTION)}. (See {@link #ERR_UNCAUGHT_EXCEPTION})
```
There might be better wording for this, esp. the first sentence. Any suggestions?

mcimadamore replied:
I think it is best to drop the first sentence, and go straight to the point, which is "The target method handle shoud not...".

Which reminds me, do we want to apply same logic we do for var handle combinators to detect if a method handle contains exceptions in its guts (by looking at signature of direct MH, or by looking for certain combinators).

JornVernee commented on test/jdk/java/foreign/TestUpcall.java:
```diff
@@ -28,7 +28,7 @@
  * @build NativeTestHelper CallGeneratorHelper TestUpcall
  *
- * @run testng/othervm
+ * @run testng/othervm/timeout=720
```
I've increased the timeout of TestUpcall here as well, as I was seeing timeouts pretty frequently

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -114,6 +115,13 @@
+     *
+     * @see CLinker#upcallStub(MethodHandle, FunctionDescriptor, ResourceScope)
+     */
+    int ERR_UNCAUGHT_EXCEPTION = privilegedGetProperty("jdk.incubator.foreign.uncaught_exception_code", 1);
```
Do we need a system property? E.g. can it be a constant value?

JornVernee replied:
We could just use `1` here as well, but during testing I've found that the default uncaught exception handler also return `1`, so I was getting a false positive.

JornVernee replied:
Thinking some more. Having a system property is probably overkill, and checking stderr for the exception message in the test is good enough

mcimadamore commented on test/jdk/java/foreign/TestUpcallException.java:
```diff
@@ -0,0 +1,83 @@
+            .start();
+
+        int result = process.waitFor();
+        assertEquals(result, exitCode);
```
Should we check for stacktrace output - e.g. the name of ThrowingUpcall" ?

JornVernee replied:
That's probably a good idea. Though, I'll note that it still allows for false positives if the tests throws an exception from ThrowingUpcall another way. Comparing the exit codes should catch that though.

mcimadamore reviewed (approved):
Looks good!

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -224,9 +232,14 @@ static SymbolLookup systemLookup() {
+     * The returned memory address is associated with the provided scope. When such scope is closed,
      * the corresponding native stub will be deallocated.
      * <p>
+     * Any exceptions that occur during an upcall should be handled during the upcall. The target method handle
```
I think it is best to drop the first sentence, and go straight to the point, which is "The target method handle shoud not...".

Which reminds me, do we want to apply same logic we do for var handle combinators to detect if a method handle contains exceptions in its guts (by looking at signature of direct MH, or by looking for certain combinators).

JornVernee replied:
I'll see about adding a check for exceptions.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java:
```diff
@@ -359,14 +362,25 @@ private static MethodHandle reachabilityFenceHandle(Class<?> type) {
         if (specializedHandle.type().returnType() == void.class) {
-            closer = empty(methodType(void.class, Throwable.class)); // (Throwable) -> void
+            if (!upcall) {
+                closer = empty(methodType(void.class, Throwable.class)); // (Throwable) -> void
```
Shouldn't ProgrammableUpcallHandler do the same thing?

JornVernee replied:
It does? Both call SharedUtils.handelUncaughtException

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java:
```diff
@@ -295,6 +295,9 @@ private static Object invokeInterpBindings(Object[] moves, MethodHandle leaf,
                 return returnMoves;
             }
+        } catch(Throwable t) {
+            SharedUtils.handleUncaughtException(t);
```
Ah - missed that :-)

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -114,6 +111,13 @@
+     *
+     * @see CLinker#upcallStub(MethodHandle, FunctionDescriptor, ResourceScope)
+     */
+    int ERR_UNCAUGHT_EXCEPTION = 1;
```
On a second thought, I'm not super sure of the value of documenting what the exit code should be? Maybe we can leave that unspecified?

JornVernee replied:
I agree. Without a system property, having a constant doesn't really make sense. I looked around in the documentation of exit codes in the JDK, and most  places seem to at least specify the exit code  to be non-zero in case of error. I'll add a note like that to the documentation of upcallStub.

mcimadamore commented on test/jdk/java/foreign/TestUpcallException.java:
```diff
@@ -0,0 +1,95 @@
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestUpcallException {
```
I wonder if we should also test with a security manager (even though that's getting deprecated) - and see what happens when you call System.exit and SM is enabled with policy to block off calls to System.exit.

JornVernee commented:
I've switched to using `Shutdown.exit`, and added a `SecurityManager` that blocks the normal `System.exit` call into the test. I've also switched `CABI` from using `System.getProperty` to determine the platform, to using `priviledgedGetProperty`, to avoid any security exceptions (found during testing).

I've removed the public `ERR_UNCAUGHT_EXCEPTION` constant, and changed the documentation to say that the VM will exit with a non-zero exit status in the case of an uncaught exception.

Finally, I've changed the test to always print the output as well, besides checking the contents.

JornVernee commented:
I've now added the eager exception checking to upcallStub.

As discussed offline, this reuses the exception check done for the VarHandle adapters, but makes the check more lenient towards BoundMetrhodHandles that might or might not throw an exception. The current check was too conservative toward rejecting such handles, because it rejects a BoundMehtodHandle that contained any MethodHandle field that could throw an exception. But, such an exception might be caught be an exclosing method handle. (see also https://bugs.openjdk.java.net/browse/JDK-8268031)

Instead, when a filter used for a var handle combinator throws a checked a exception when invoked, it is caught, and an IllegalStateException is thrown instead (with the original exception as the cause).

JornVernee commented on src/java.base/share/classes/java/lang/invoke/VarHandles.java:
```diff
@@ -604,42 +600,55 @@ public static VarHandle dropCoordinates(VarHandle target, int pos, Class<?>... v
+            return new Class<?>[0];
         }
+
+        assert handle instanceof BoundMethodHandle : "Unecpexted handle type: " + handle;
```
```suggestion
        assert handle instanceof BoundMethodHandle : "Unexpected handle type: " + handle;
```

mcimadamore reviewed (approved):
Looks very good - thanks for fixing the VH adaptation code as well!

JornVernee commented:
/solves 8268031

JornVernee commented:
/integrate
